File: examples/python/advanced_usage.py
# Example: Advanced Usage with Error Handling

import asyncio
import logging
from synora import SynoraClient, SynoraError, PaymentRequiredError, PaymentFailedError

# Setup logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


async def advanced_example():
    """Advanced usage with error handling and retry logic"""
    
    client = SynoraClient(
        wallet_private_key="0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef",
        max_retries=3,
        timeout=30
    )
    
    # Check balance first
    try:
        balance = await client.get_balance()
        logger.info(f"Current balance: ${balance}")
        
        if balance < 0.10:
            logger.warning("Low balance! Please add more USDC")
            return
    except Exception as e:
        logger.error(f"Failed to check balance: {e}")
    
    # Multiple API calls with error handling
    tasks = [
        ('analyze', '/tools/analyze', {'text': 'Hello world', 'model': 'gpt-4'}),
        ('ocr', '/tools/ocr', {'image_url': 'https://example.com/image.png'}),
        ('search', '/tools/search', {'query': 'Synora', 'max_results': 3}),
    ]
    
    for name, endpoint, data in tasks:
        try:
            logger.info(f"Calling {name}...")
            result = await client.call(endpoint, data)
            logger.info(f"{name} success: {result}")
            
        except PaymentRequiredError as e:
            logger.error(f"Payment required for {name}: {e.payment_details}")
            
        except PaymentFailedError as e:
            logger.error(f"Payment failed for {name}: {e}")
            
        except SynoraError as e:
            logger.error(f"Synora error for {name}: {e}")
            
        except Exception as e:
            logger.error(f"Unexpected error for {name}: {e}")
        
        # Small delay between requests
        await asyncio.sleep(0.5)


if __name__ == '__main__':
    asyncio.run(advanced_example())

File: examples/javascript/advanced_usage.ts
// Example: Advanced Usage with Error Handling

import { SynoraClient, SynoraError, PaymentRequiredError, PaymentFailedError } from '@synora/sdk';

async function advancedExample() {
  const client = new SynoraClient({
    walletPrivateKey: '0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef',
    maxRetries: 3,
    timeout: 30000
  });

  // Check balance first
  try {
    const balance = await client.getBalance();
    console.log(`Current balance: $${balance}`);

    if (balance < 0.10) {
      console.warn('Low balance! Please add more USDC');
      return;
    }
  } catch (error) {
    console.error('Failed to check balance:', error);
  }

  // Multiple API calls with error handling
  const tasks = [
    { name: 'analyze', endpoint: '/tools/analyze', data: { text: 'Hello world', model: 'gpt-4' } },
    { name: 'ocr', endpoint: '/tools/ocr', data: { image_url: 'https://example.com/image.png' } },
    { name: 'search', endpoint: '/tools/search', data: { query: 'Synora', max_results: 3 } }
  ];

  for (const task of tasks) {
    try {
      console.log(`Calling ${task.name}...`);
      const result = await client.call(task.endpoint, task.data);
      console.log(`${task.name} success:`, result);

    } catch (error) {
      if (error instanceof PaymentRequiredError) {
        console.error(`Payment required for ${task.name}:`, error.paymentDetails);
      } else if (error instanceof PaymentFailedError) {
        console.error(`Payment failed for ${task.name}:`, error.message);
      } else if (error instanceof SynoraError) {
        console.error(`Synora error for ${task.name}:`, error.message);
      } else {
        console.error(`Unexpected error for ${task.name}:`, error);
      }
    }

    // Small delay between requests
    await new Promise(resolve => setTimeout(resolve, 500));
  }
}

advancedExample().catch(console.error);